### PR TITLE
Geneformer gene embeddings calculation. Now limits the changes to bionemo.geneformer only

### PR DESCRIPTION
**Pull request by jyin-bst**
### Type of changes

- [ ]  Bug fix (non-breaking change which fixes an issue)
- [x ]  New feature (non-breaking change which adds functionality)
- [ ]  Refactor
- [ ]  Documentation update
- [ ]  Other (please describe):

### CI Pipeline Configuration
Configure CI behavior by applying the relevant labels:

- [SKIP_CI](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/user-guide/contributing/contributing.md#skip_ci) - Skip all continuous integration tests
- [INCLUDE_NOTEBOOKS_TESTS](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/user-guide/contributing/contributing.md#include_notebooks_tests) - Execute notebook validation tests in pytest
- [INCLUDE_SLOW_TESTS](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/user-guide/contributing/contributing.md#include_slow_tests) - Execute tests labelled as slow in pytest for extensive testing

> [!NOTE]
> By default, the notebooks validation tests are skipped unless explicitly enabled.

#### Authorizing CI Runs

We use [copy-pr-bot](https://docs.gha-runners.nvidia.com/apps/copy-pr-bot/#automation) to manage authorization of CI
runs on NVIDIA's compute resources.

* If a pull request is opened by a trusted user and contains only trusted changes, the pull request's code will
  automatically be copied to a pull-request/ prefixed branch in the source repository (e.g. pull-request/123)
* If a pull request is opened by an untrusted user or contains untrusted changes, an NVIDIA org member must leave an
  `/ok to test` comment on the pull request to trigger CI. This will need to be done for each new commit.

### Usage

```python
TODO: Add code snippet
```

### Pre-submit Checklist

 - [x] I have tested these changes locally
 - [ ] I have updated the documentation accordingly
 - [ ] I have added/updated tests as needed
 - [ ] All existing tests pass successfully

### CODE REVIEW

File: sub-packages/bionemo-geneformer/src/bionemo/geneformer/utils/callbacks.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+        batch_indices: Sequence[int],
+    ) -> None:
+        """Writes predictions to disk at the end of each epoch.
+
+        Args:
+            trainer: The Trainer instance.
+            pl_module: The LightningModule instance.
+            predictions: The predictions made by the model.
+            batch_indices: The indices of the batch.
+        """
```
Comment (skothenhill-nv) [inline sub-packages/bionemo-geneformer/src/bionemo/geneformer/utils/callbacks.py:101]: Could you add an example of what the output may contain? We had discussed potentially empty gene-ids for unobserved gene tokens, please include this information in the docstring as well.

`batch_indices` is unused, we should either remove it, replace it with **kwargs (which will do nothing), or leave as-is and expand the docstring to state its unused. I'd prefer to remove it, but if there is an API conflict we can replace it with **kwargs.

`pl_module` same comment applies here, seems unused, so we could collapse this into *args or atleast state its unused.

`trainer`: state its used for getting global rank

`predictions` I think is good as is.

File: sub-packages/bionemo-geneformer/src/bionemo/geneformer/utils/callbacks.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+            for i in range(cell_count):
+                cell_state = hidden_states[i]
+                cell_input_ids = input_ids[i].cpu().numpy()
+
+                # iterate over each gene in the cell
+                for idx, embedding in zip(cell_input_ids, cell_state):
+
+                    # idx < 5 are dummy tokens based on geneformer.vocab
+                    if idx < 5:
+                        continue
```
Comment (skothenhill-nv) [inline sub-packages/bionemo-geneformer/src/bionemo/geneformer/utils/callbacks.py:142]: Could we infer this from the tokenizer rather than using a magic number?

File: sub-packages/bionemo-geneformer/src/bionemo/geneformer/utils/callbacks.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+                        # increment input_id count
+                        input_id_count[idx] += 1
+
+            # divide each embedding sum by the total occurences of each gene to get an average
+            for input_id in gene_embedding_accumulator.keys():
+                gene_embedding_accumulator[input_id] /= input_id_count[input_id]
+
+            logging.info("Finished calculating gene embeddings.")
+
+            prediction['gene_embeddings'] = gene_embedding_accumulator
```
Comment (skothenhill-nv) [inline sub-packages/bionemo-geneformer/src/bionemo/geneformer/utils/callbacks.py:164]: Would it be helpful to also return the `input_id_count`? Could help with quality filtering or something- if a gene token only has one observation it could warp downstream analyses? What do you think?

**Review by skothenhill-nv — CHANGES_REQUESTED** [review]
Left mostly style comments, this work needs accompanying tests as well.

**Comment by jyin-bst** [comment]
> Left mostly style comments, this work needs accompanying tests as well.

Thank you so much Steven. I will look into these suggestions and reply to you in the next few days.

File: sub-packages/bionemo-geneformer/src/bionemo/geneformer/api.py [hunk 4]
```diff
@@ -71,6 +71,8 @@ class GeneformerConfig(BioBertConfig[GeneformerModel, MegatronLossType], iom.IOM
     get_attention_mask_from_fusion: bool = True
 
     position_embedding_type: PositionEmbeddingKinds = "learned_absolute"
+    include_gene_embeddings: bool = False
+
```
[the remark this answers is not in the document]
Comment (jyin-bst) [reply sub-packages/bionemo-geneformer/src/bionemo/geneformer/api.py:75]: This has to be included, otherwise, it will give "got an unexpected keyword argument 'include_gene_embeddings'" error

File: sub-packages/bionemo-geneformer/src/bionemo/geneformer/utils/callbacks.py (hunk 1 above)
Comment (jyin-bst) [reply sub-packages/bionemo-geneformer/src/bionemo/geneformer/utils/callbacks.py:101]: Yes, the description was copied from sub-packages/bionemo-llm/src/bionemo/llm/utils/callbacks.py. I will see whether we can remove them.

**Comment by isabel-wilkinson** [comment]
As discussed in the call today, @jyin-bst will prepare a notebook tutorial demonstrating use of the gene embedding extraction which will double as an integration test.
This tutorial notebook should be placed in bionemo-framework/sub-packages/bionemo-geneformer/examples/
@jstjohn to help with turning the notebook into a brev.dev launchable.

File: sub-packages/bionemo-geneformer/src/bionemo/geneformer/api.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (skothenhill-nv) [reply sub-packages/bionemo-geneformer/src/bionemo/geneformer/api.py:75]: wouldn't that be because it's passed into `config_class`? Which is my concern, its not related to the config, not used by the config, but we pass it in. It's used only by the inference script. @isabel-wilkinson could you take a look?

[the remark this answers is not in the document]
Comment (isabel-wilkinson) [reply sub-packages/bionemo-geneformer/src/bionemo/geneformer/api.py]: yes i remove this now

File: sub-packages/bionemo-geneformer/src/bionemo/geneformer/scripts/infer_geneformer.py [hunk 5]
```diff
@@ -131,11 +133,13 @@ def infer_model(
         # handle checkpoint resumption here rather than auto-resume so this supports fine-tuning capabilities
         initial_ckpt_path=str(checkpoint_path) if checkpoint_path is not None else None,
         include_embeddings=include_embeddings,
+        include_gene_embeddings=include_gene_embeddings,
```
[the remark this answers is not in the document]
Comment (isabel-wilkinson) [reply sub-packages/bionemo-geneformer/src/bionemo/geneformer/scripts/infer_geneformer.py]: removed this now

File: sub-packages/bionemo-geneformer/src/bionemo/geneformer/utils/callbacks.py (hunk 2 above)
Comment (skothenhill-nv) [reply sub-packages/bionemo-geneformer/src/bionemo/geneformer/utils/callbacks.py]: sub-packages/bionemo-geneformer/src/bionemo/geneformer/tokenizer/gene_tokenizer.py:L92 I think is relevant- returns all 'special' tokens- we then can infer the remaining are valid gene tokens.

@isabel-wilkinson

File: sub-packages/bionemo-geneformer/src/bionemo/geneformer/utils/callbacks.py (hunk 3 above)
Comment (isabel-wilkinson) [reply sub-packages/bionemo-geneformer/src/bionemo/geneformer/utils/callbacks.py]: added

File: sub-packages/bionemo-geneformer/src/bionemo/geneformer/utils/callbacks.py (hunk 2 above)
Comment (isabel-wilkinson) [reply sub-packages/bionemo-geneformer/src/bionemo/geneformer/utils/callbacks.py]: done

**Review by jstjohn — APPROVED** [review]
I added the slow test tag to make sure this doesn't require any other test changes, but otherwise looks good. Thanks!

**Comment by isabel-wilkinson** [comment]
/ok to test https://github.com/NVIDIA/bionemo-framework/pull/808/commits/3fe6e9293486f6a19d248d6b23a3ac39a406eddd

**Comment by isabel-wilkinson** [comment]
> I added the slow test tag to make sure this doesn't require any other test changes, but otherwise looks good. Thanks!

Hi @jstjohn I think the slow test tag results in test failure due to failures in the `discrete_data_interpolant_tutorial.ipynb` notebook from moco. It seems to be non deterministic - I first had failure in cell 23, and after re-running I got a failure in cell 4. I cannot reproduce this issue locally on my RTX 5880 Ada GPU. I believe this is a known issue and unrelated to my PR: https://nvidia.slack.com/archives/C074Z808N05/p1737049409526399
It seems to me like the slow tests tag was introduced as a result of this issue to allow skipping of running old notebooks.

The other notebooks all pass in the CI test.

Can I remove this slow test tag or what do you recommend I do to proceed here?
